### PR TITLE
Add parent field to records to represent repeat and retry information.

### PR DESCRIPTION
**Pull request by booneng**
This change is [Reviewable](https://reviewable.io/reviews/google/mobly/905)

### CODE REVIEW

File: mobly/records.py [hunk 1]
```diff
@@ -320,6 +328,9 @@ class TestResultRecord:
     retry_parent: TestResultRecord, only set for retry iterations. This is the
```
Comment (xpconanfan) [inline mobly/records.py:328]: mark this as deprecate?

File: mobly/base_test.py [hunk 2]
```diff
@@ -720,6 +720,7 @@ def should_retry(record):
       retry_name = f'{test_name}_retry_{i+1}'
```
Comment (xpconanfan) [inline mobly/base_test.py:720]: Trying to discuss here.

Should we still generate new names, now that the record clearly marks the sequence.
I guess it's easy for humans to understand.
Does the downstream result process also try to add the postfixes?

**Review by xpconanfan — CHANGES_REQUESTED** [review]

**Review by booneng — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/google/mobly/905)* status: 0 of 4 files reviewed, 2 unresolved discussions (waiting on @booneng and @xpconanfan)

___
*[`mobly/base_test.py` line 720 at r1](https://reviewable.io/reviews/google/mobly/905#gh-1436114200:-Nm_1gt4BHFG7zOP75q6:b-lhied) ([raw file](https://github.com/google/mobly/blob/86803fe0f812527a89dc2e02d58cd72144a29ec1/mobly/base_test.py#L720)):*
<details><summary><i>Previously, xpconanfan (Ang Li) wrote…</i></summary><blockquote>

Trying to discuss here.

Should we still generate new names, now that the record clearly marks the sequence.
I guess it's easy for humans to understand.
Does the downstream result process also try to add the postfixes?
</blockquote></details>

MH adds their own suffix but we are trying to fix that. Otherwise, no other process adds suffixes. I think we should remove it in another change because other processes read the suffixes to determine the repeat/retry number.
___
*[`mobly/records.py` line 328 at r1](https://reviewable.io/reviews/google/mobly/905#gh-1436113803:-Nm_1g3rCyDgwD1-sIov:b-896fix) ([raw file](https://github.com/google/mobly/blob/86803fe0f812527a89dc2e02d58cd72144a29ec1/mobly/records.py#L328)):*
<details><summary><i>Previously, xpconanfan (Ang Li) wrote…</i></summary><blockquote>

mark this as deprecate?
</blockquote></details>

Done.

**Review by xpconanfan — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/google/mobly/905)* status: 0 of 4 files reviewed, 1 unresolved discussion (waiting on @booneng)

___
*[`mobly/records.py` line 328 at r1](https://reviewable.io/reviews/google/mobly/905#gh-1436113803:-NmbBGV3Diyg3ZeQwAP2:b-mxsjnh) ([raw file](https://github.com/google/mobly/blob/86803fe0f812527a89dc2e02d58cd72144a29ec1/mobly/records.py#L328)):*
<details><summary><i>Previously, booneng (boon) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

should probably mention it's deprecated at the very beginning of the sentence, like `[DEPRECATED] Use ... instead.` before the sentence starts.
or is there an official way of marking deprecation in sphinx, our doc generator?

**Review by booneng — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/google/mobly/905)* status: 0 of 4 files reviewed, 1 unresolved discussion (waiting on @booneng and @xpconanfan)

___
*[`mobly/records.py` line 328 at r1](https://reviewable.io/reviews/google/mobly/905#gh-1436113803:-Nmdtldx-Jg6zJ-reTwN:b-896fix) ([raw file](https://github.com/google/mobly/blob/86803fe0f812527a89dc2e02d58cd72144a29ec1/mobly/records.py#L328)):*
<details><summary><i>Previously, xpconanfan (Ang Li) wrote…</i></summary><blockquote>

should probably mention it's deprecated at the very beginning of the sentence, like `[DEPRECATED] Use ... instead.` before the sentence starts.
or is there an official way of marking deprecation in sphinx, our doc generator?
</blockquote></details>

Done.

**Review by xpconanfan — APPROVED** [review]

**Pull request merged**